components/msg_typing/msg_typing.tsx

```diff
@@ -1,19 +1,47 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-import React from 'react';
+import React, {useCallback} from 'react';
 import {FormattedMessage} from 'react-intl';
 
+import {WebSocketMessage} from '@mattermost/client';
+
+import {SocketEvents} from 'utils/constants';
+import {useWebSocket} from 'utils/use_websocket';
+
 type Props = {
     typingUsers: string[];
+    userStartedTyping: (userId: string, channelId: string, rootId: string, now: number) => void;
+    userStoppedTyping: (userId: string, channelId: string, rootId: string, now: number) => void;
 }
 
-export default class MsgTyping extends React.PureComponent<Props> {
-    private getTypingText = () => {
+export default function MsgTyping(props: Props) {
+    const {userStartedTyping, userStoppedTyping} = props;
+    useWebSocket({
+        handler: useCallback((msg: WebSocketMessage) => {
+            if (msg.event === SocketEvents.TYPING) {
+                const channelId = msg.broadcast.channel_id;
+                const rootId = msg.data.parent_id;
+                const userId = msg.data.user_id;
+
+                userStartedTyping(userId, channelId, rootId, Date.now());
```

Comment:
Why is this calling out to Redux? Won't this cause a redux event for every instance of the component? (I guess it's max 2 but still) Shouldn't we handle this state locally? (custom useTyping hook?)

Reply:
Good catch. I'll change this to check the channel ID and root ID first to avoid doubling up on redux events.

You're right that this state could probably be stored locally though. I had tried that as part of my hackathon project, but something about storing a Set in state was causing me problems, so I abandoned it. That'd definitely be a good improvement down the line since it'd get rid of some redux state, but it's out of scope for what I want to change here